### PR TITLE
feat(rln): add recover identity secret API

### PR DESCRIPTION
**Pull request by s1fr0**
This PR adds public and FFI API to recover the identity secret from two RLN proofs generated for same secret within same epoch, i.e. when spam is detected. The recovered secret can then be used for slashing.

Tests are added too.

### CODE REVIEW

File: rln/src/protocol.rs [hunk 1]
```diff
@@ -426,6 +427,30 @@ pub fn hash_to_field(signal: &[u8]) -> Fr {
...
+    let (x1, y1) = share1;
+    let (x2, y2) = share2;
+
+    // If the two input shares were computed for the same epoch and identity secret, we can recover the latter
+    // y1 = a_0 + x1 * a_1
+    // y2 = a_0 + x2 * a_1
+    let a_1 = (y1 - y2) / (x1 - x2);
+    let a_0 = y1 - x1 * a_1;
+
+    // If shares come the same polynomial, a0 is correctly recovered and a1 = poseidonHash([a0, epoch])
```
Comment (oskarth) [inline rln/src/protocol.rs:443]:
```suggestion
    // If shares come from the same polynomial, a0 is correctly recovered and a1 = poseidonHash([a0, epoch])
```

File: rln/src/public.rs [hunk 2]
```diff
@@ -528,7 +528,7 @@ impl RLN<'_> {
     /// rln.generate_rln_proof(&mut input_buffer, &mut output_buffer)
     ///     .unwrap();
     ///
-    /// // proof_data is [ proof<128> | share_y<32> | nullifier<32> | root<32> | epoch<32> | share_x<32> | rln_identifier<32> ]
+    /// // proof_data is [ proof<128> | root<32> | epoch<32> | share_x<32> | share_y<32> | nullifier<32> |  rln_identifier<32> ]
```
Comment (oskarth) [inline rln/src/public.rs:531]: Is this just a typo fix?

**Review by oskarth — APPROVED** [review]
Nice one!

File: rln/src/public.rs [hunk 3]
```diff
@@ -830,6 +830,77 @@ impl RLN<'_> {
...
+            let recovered_id_secret = compute_id_secret(share1, share2, proof_values_1.epoch);
+
+            // If an id secret is recovered, we write it to output_data, otherwise nothing will be written.
+            if recovered_id_secret.is_ok() {
+                let id_secret = recovered_id_secret.unwrap();
+                output_data.write_all(&fr_to_bytes_le(&id_secret))?;
+            }
+        }
+
+        return Ok(());
```
Comment (rymnc) [inline rln/src/public.rs:901]:
```suggestion
        Ok(())
```
Based on other functions in this module :)

**Review by rymnc — COMMENTED** [review]
In general LGTM, minor nits :)

**Comment by oskarth** [comment]
@s1fr0 could we merge this today? I'd like to use for workshop tomorrow morning

File: rln/src/public.rs [hunk 4]
```diff
@@ -1619,4 +1690,144 @@ mod test {
...
+        let serialized_id_secret = output_buffer.into_inner();
+
+        // We ensure that a non-empty value is written to output_buffer
+        assert!(!serialized_id_secret.is_empty());
+
+        // We check if the recovered id secret corresponds to the original one
+        let (recovered_id_secret, _) = bytes_le_to_fr(&serialized_id_secret);
+        assert_eq!(recovered_id_secret, identity_secret);
+
+        // We now test that computing_id_secret is unsuccessful if shares computed from two different id secrets but within same epoch are passed
```
[the remark this answers is not in the document]
Comment (rymnc) [reply rln/src/public.rs:1774]: Can be addressed in a follow up PR, then. In general we should be testing each branch of the code, but in different tests for better granularity :)

**Review by rymnc — APPROVED** [review]
LGTM

**Pull request merged**